File: web/src/retrieval.ts
/**
 * TypeScript port of apps/prototype-cli/retrieval.py.
 *
 * Fuzzy/keyword matching of a typed question against the local item bank.
 * No LLM, no network calls — pure client-side logic, subject-agnostic (it
 * only ever looks at question_text, so it works for biology, chemistry, or
 * any future subject folder without changes).
 */
import type { Item } from "./types";

const STOPWORDS = new Set([
  "a", "an", "the", "is", "are", "was", "were", "of", "in", "on", "at",
  "to", "for", "and", "or", "which", "what", "who", "whom", "does",
  "do", "did", "how", "why", "when", "where", "called", "known", "as",
  "this", "that", "these", "those", "it", "its", "with", "by", "term",
  "describes",
]);

// Matches Python's string.punctuation.
const PUNCTUATION_RE = /[!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~]/g;

function normalize(text: string): string {
  return text.toLowerCase().replace(PUNCTUATION_RE, "");
}

function keywords(text: string): Set<string> {
  const words = normalize(text)
    .split(/\s+/)
    .filter((word) => word.length > 0 && !STOPWORDS.has(word));
  return new Set(words);
}

/**
 * Ratcliff/Obershelp ratio, matching Python's difflib.SequenceMatcher.ratio()
 * (autojunk is not implemented — Python only applies it to strings >= 200
 * chars with a popular single character, which never applies to question
 * text or single keywords here).
 */
function findLongestMatch(
  a: string,
  b: string,
  alo: number,
  ahi: number,
  blo: number,
  bhi: number,
  b2j: Map<string, number[]>
): { a: number; b: number; size: number } {
  let besti = alo;
  let bestj = blo;
  let bestsize = 0;
  let j2len = new Map<number, number>();

  for (let i = alo; i < ahi; i++) {
    const newj2len = new Map<number, number>();
    const indices = b2j.get(a[i]) ?? [];
    for (const j of indices) {
      if (j < blo) continue;
      if (j >= bhi) break;
      const k = (j2len.get(j - 1) ?? 0) + 1;
      newj2len.set(j, k);
      if (k > bestsize) {
        besti = i - k + 1;
        bestj = j - k + 1;
        bestsize = k;
      }
    }
    j2len = newj2len;
  }

  return { a: besti, b: bestj, size: bestsize };
}

function getMatchingBlocks(a: string, b: string): Array<[number, number, number]> {
  const b2j = new Map<string, number[]>();
  for (let j = 0; j < b.length; j++) {
    const c = b[j];
    const list = b2j.get(c);
    if (list) list.push(j);
    else b2j.set(c, [j]);
  }

  const queue: Array<[number, number, number, number]> = [[0, a.length, 0, b.length]];
  const blocks: Array<[number, number, number]> = [];

  while (queue.length > 0) {
    const [alo, ahi, blo, bhi] = queue.pop()!;
    const { a: i, b: j, size: k } = findLongestMatch(a, b, alo, ahi, blo, bhi, b2j);
    if (k > 0) {
      blocks.push([i, j, k]);
      if (alo < i && blo < j) queue.push([alo, i, blo, j]);
      if (i + k < ahi && j + k < bhi) queue.push([i + k, ahi, j + k, bhi]);
    }
  }

  blocks.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  return blocks;
}

export function sequenceMatcherRatio(a: string, b: string): number {
  const total = a.length + b.length;
  if (total === 0) return 1.0;
  const blocks = getMatchingBlocks(a, b);
  let matches = 0;
  for (const [, , size] of blocks) matches += size;
  return (2.0 * matches) / total;
}

/** Jaccard-style overlap where words count as equal if they're close enough
 * (handles typos like 'organel'/'organelle', 'cel'/'cell'). */
function fuzzyKeywordOverlap(
  queryKw: Set<string>,
  itemKw: Set<string>,
  threshold = 0.75
): number {
  if (queryKw.size === 0 || itemKw.size === 0) return 0.0;

  const remaining = new Set(itemKw);
  let matches = 0;
  for (const qw of queryKw) {
    let bestWord: string | null = null;
    let bestRatio = 0.0;
    for (const iw of remaining) {
      const ratio = sequenceMatcherRatio(qw, iw);
      if (ratio > bestRatio) {
        bestWord = iw;
        bestRatio = ratio;
      }
    }
    if (bestRatio >= threshold && bestWord !== null) {
      matches += 1;
      remaining.delete(bestWord);
    }
  }

  const union = queryKw.size + itemKw.size - matches;
  return union > 0 ? matches / union : 0.0;
}

function score(query: string, item: Item): number {
  const queryNorm = normalize(query);
  const itemNorm = normalize(item.question_text);
  const seqRatio = sequenceMatcherRatio(queryNorm, itemNorm);

  const queryKw = keywords(query);
  const itemKw = keywords(item.question_text);
  const overlap = fuzzyKeywordOverlap(queryKw, itemKw);

  return 0.3 * seqRatio + 0.7 * overlap;
}

export interface MatchResult {
  item: Item | null;
  confidence: number;
}

/** Return the best-matching item and a 0.0-1.0 confidence score. Never
 * refuses on its own — the caller (refusal.ts) decides what to do with a
 * low score. */
export function match(query: string, items: Item[]): MatchResult {
  if (items.length === 0 || query.trim().length === 0) {
    return { item: null, confidence: 0.0 };
  }

  let bestItem: Item | null = null;
  let bestScore = 0.0;
  for (const item of items) {
    const s = score(query, item);
    if (s > bestScore) {
      bestItem = item;
      bestScore = s;
    }
  }

  return { item: bestItem, confidence: Math.round(bestScore * 10000) / 10000 };
}

File: apps/prototype-cli/retrieval.py
"""Fuzzy/keyword matching of a typed question against the local item bank.

No LLM, no external API calls — stdlib only. v0 scope: NEET Biology, text only.
"""
import json
import string
from difflib import SequenceMatcher
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
DEFAULT_ITEM_BANK_GLOB = "item_bank/biology/*.jsonl"

_STOPWORDS = {
    "a", "an", "the", "is", "are", "was", "were", "of", "in", "on", "at",
    "to", "for", "and", "or", "which", "what", "who", "whom", "does",
    "do", "did", "how", "why", "when", "where", "called", "known", "as",
    "this", "that", "these", "those", "it", "its", "with", "by", "term",
    "describes",
}


def _normalize(text):
    text = text.lower()
    text = text.translate(str.maketrans("", "", string.punctuation))
    return text


def _keywords(text):
    return {word for word in _normalize(text).split() if word and word not in _STOPWORDS}


def load_items(glob_pattern=DEFAULT_ITEM_BANK_GLOB, root=REPO_ROOT):
    """Load every item from *.jsonl files matching glob_pattern under root."""
    items = []
    for path in sorted(root.glob(glob_pattern)):
        with path.open(encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                items.append(json.loads(line))
    return items


def _fuzzy_keyword_overlap(query_kw, item_kw, threshold=0.75):
    """Jaccard-style overlap where words count as equal if they're close
    enough (handles typos like 'organel'/'organelle', 'cel'/'cell')."""
    if not query_kw or not item_kw:
        return 0.0

    remaining = set(item_kw)
    matches = 0
    for qw in query_kw:
        best_word, best_ratio = None, 0.0
        for iw in remaining:
            ratio = SequenceMatcher(None, qw, iw).ratio()
            if ratio > best_ratio:
                best_word, best_ratio = iw, ratio
        if best_ratio >= threshold:
            matches += 1
            remaining.discard(best_word)

    union = len(query_kw) + len(item_kw) - matches
    return matches / union if union else 0.0


def _score(query, item):
    query_norm = _normalize(query)
    item_norm = _normalize(item["question_text"])
    seq_ratio = SequenceMatcher(None, query_norm, item_norm).ratio()

    query_kw = _keywords(query)
    item_kw = _keywords(item["question_text"])
    overlap = _fuzzy_keyword_overlap(query_kw, item_kw)

    return 0.3 * seq_ratio + 0.7 * overlap


def match(query, items=None):
    """Return (best_item_or_None, confidence) for query against items.

    confidence is 0.0-1.0. Caller decides what to do with a low score —
    this function never refuses, it only scores.
    """
    if items is None:
        items = load_items()
    if not items or not query.strip():
        return None, 0.0

    best_item, best_score = None, 0.0
    for item in items:
        score = _score(query, item)
        if score > best_score:
            best_item, best_score = item, score

    return best_item, round(best_score, 4)
